spring-kafka/src/main/java/org/springframework/kafka/core/DefaultTransactionSuffixManager.java

```diff
@@ -0,0 +1,110 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.core;
+
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * The {@link TransactionSuffixManager} implementation for managing transactional producer suffixes.
+ * If the {@link #maxCache} is greater than 0, the suffixes will be cached and reused.
+ * Otherwise, the suffixes will be generated on the fly.
+ *
+ * <p>
+ * Setting the {@link #setMaxCache(int)} enables caching to restrict the number of `transactional.id`.
+ * The cache is per `transactional.id` prefix. The cache size is limited by the {@link #maxCache}.
+ * Default is 0, which means no caching and restriction, so the `transactional.id` will be generated on the fly.
+ *
+ * @author Ilya Starchenko
+ *
+ * @since 3.1.1
+ */
+public class DefaultTransactionSuffixManager implements TransactionSuffixManager {
+
+	private final AtomicInteger transactionIdSuffix = new AtomicInteger();
+
+	private final Map<String, BlockingQueue<String>> suffixCache = new ConcurrentHashMap<>();
+
+	private int maxCache;
+
+	/**
+	 * Retrieve the suffix for the transactional producer from the cache or generate a new one 
+	 * if caching is disabled.
+	 * @param txIdPrefix the transaction id prefix.
+	 * @return the suffix.
+	 * @throws NoProducerAvailableException if caching is enabled and no suffixes are available.
+	 */
+	@Override
+	public String retrieveSuffix(String txIdPrefix) {
+		BlockingQueue<String> cache = getSuffixCache(txIdPrefix);
+		if (cache == null) {
+			return String.valueOf(this.transactionIdSuffix.getAndIncrement());
+		}
+
+		String suffix = cache.poll();
+		if (suffix == null) {
+			throw new NoProducerAvailableException("No available transaction producer suffix for " + txIdPrefix);
+		}
+		return suffix;
+	}
+
+	@Override
+	public void returnSuffix(@Nullable String txIdPrefix, @Nullable String suffix) {
+		if (this.maxCache <= 0 || suffix == null) {
+			return;
+		}
+		BlockingQueue<String> queue = getSuffixCache(txIdPrefix);
+		if (queue != null && !queue.contains(suffix)) {
+			queue.add(suffix);
+		}
+	}
+
+	@Override
+	public void reset() {
+		this.suffixCache.clear();
+	}
+
+	/**
+	 * Set the maximum size for transaction producer cache.
+	 * @param maxCache the maxCache to set
+	 */
+	public void setMaxCache(int maxCache) {
+		Assert.isTrue(maxCache >= 0, "'maxCache' must greater than or equal 0");
+		this.maxCache = maxCache;
+	}
+
+	@Nullable
+	private BlockingQueue<String> getSuffixCache(@Nullable String txIdPrefix) {
+		if (txIdPrefix == null || this.maxCache <= 0) {
```

Comment:
When txIdPrefix is null maybe not return suffix is better.
At DefaultKafkaProducerFactory line 848 in this PR, When txIdPrefix is null will threw exception.

Reply:
Sounds reasonable; I will add an assert.

Reply:
Fixed.